src/Language/PureScript/Scheme/CodeGen/Transpiler.hs

```diff
@@ -140,8 +161,23 @@ caseToScheme values caseAlternatives =
     -- We have to replace each occurrence of `m' with `v' and each occurrence
     -- of `n' with `v0'.
     variablesToReplace valuesAndBinders =
-      [ (runIdent ident, exprToScheme value)
-      | (value, (VarBinder _ann ident)) <- valuesAndBinders ]
+      concatMap go valuesAndBinders
+      where
+        go (_value, NullBinder _ann) = []
+        go (_value, LiteralBinder _ann _literal) = []
+        go (value, VarBinder _ann ident) = [(runIdent ident, exprToScheme value)]
+        go (value, ConstructorBinder _ann _typeName _constructorName binders) =
+          map (\(i, b) -> (go' b, vectorRef (exprToScheme value)
+                                            (IntegerLiteral i)))
+              (zip [1 .. (genericLength binders)]
+                   binders)
+        go _ = error "Not implemented"
+
+        -- TODO: possible fire hazard. So far I've met no ConstructorBinder
+        -- whose binders are anything but VarBinders. But if that doesn't hold
+        -- then it will explode here.
+        go' (VarBinder _ann ident) = runIdent ident
+        go' _ = error "Not implemented"
```

Comment:
Does this break it?

```
data Foo = Foo String

data Bar = Bar Foo

match :: Bar -> Int
match (Bar (Foo "")) = 1
match _ = 23
```

Reply:
I can't really tell because strings are not implemented yet.

But I can test this code (basically the same swapping String with Int):

```
data Foo = Foo Int

data Bar = Bar Foo

match :: Bar -> Int
match (Bar (Foo 0)) = 1
match _ = 23
```

And it results into:

```
(define Foo
  (lambda (value0)
    (cons (quote Foo)
              (vector value0))))

(define Bar
  (lambda (value0)
    (cons (quote Bar)
              (vector value0))))

(define match
  (lambda (v)
    (cond ((and (eq? (car v) (quote Bar))
                (eq? (car (vector-ref v 1)) (quote Foo))
                (= (vector-ref (vector-ref v 1) 1) 0)) 1)
          (#t 23))))
```

It looks correct to me, the only bug being the off-by-one error we already know about.

Reply:
Just found another bug: we should run `vector-ref` against the `cdr` of `v` and not `v` itself.

Reply:
It's definitely time to run this generated Scheme files

Reply:
Let's open a new card